cypress/integration/unit_tests/authenticate_test.js

```diff
@@ -0,0 +1,214 @@
+  const primaryPath = eeLegacyPrefix + getScoreAssetPath();
+  // Avoid non-determinism of backup asset existing by making sure it doesn't.
+  cy.stub(Resources, 'getBackupScoreAssetPath').returns('does/not/exist');
+  const {publicPolicy} = getPolicies();
+
+  let getResolveFunction;
+  const getPromise = new Promise((resolve) => getResolveFunction = resolve);
+  const getStub = cy.stub(ee.data, 'getIamPolicy').callsFake(() => {
+    getResolveFunction();
+    return Promise.resolve(publicPolicy);
+  });
+  const setIamStub = cy.stub(ee.data, 'setIamPolicy');
+  waitForTrackAndAssertNormalStubs();
+  cy.wrap(getPromise).then(() => {
+    expect(getStub).to.be.calledOnce;
+    expect(getStub).to.be.calledWith(primaryPath);
+    expect(setIamStub).to.not.be.called;
+  });
+});
+
+/**
+ * Reaches deep into EarthEngine internals and constructs IamPolicy objects for
+ * returning in {@link ee.data.getIamPolicy} stubs. Calling that method doesn't
+ * seem to work in tests, even for assets we normally have write access to.
+ *
+ * This is so gross, but it allows us to construct genuine Policy objects so
+ * that we know our code can deal with them. If names change, test will break,
+ * but should be able to find new ones.
+ * TODO(janakr): See what EE says about how to better construct policies.
+ *
+ * @return {Object} Collection of policies for use in tests
+ */
+function getPolicies() {
+  const PolicyConstructor = module$exports$eeapiclient$ee_api_client.Policy;
+  const BindingConstructor = module$exports$eeapiclient$ee_api_client.Binding;
+  const ownerBinding = new BindingConstructor({
+    role: 'roles/owner',
+    members: ['user:gd-earthengine-user@givedirectly.org'],
+  });
+  const privateReaderBinding = new BindingConstructor({
+    role: 'roles/viewer',
+    members: ['user:gd-earthengine-user@givedirectly.org'],
+  });
+  const allReadWithReaderBinding = new BindingConstructor({
+    role: 'roles/viewer',
+    members: ['user:gd-earthengine-user@givedirectly.org', 'allUsers'],
+  });
+  const allReadBinding =
+      new BindingConstructor({role: 'roles/viewer', members: ['allUsers']});
+  return {
+    noReadPolicy: new PolicyConstructor({bindings: [ownerBinding]}),
+    privateReadPolicy:
+        new PolicyConstructor({bindings: [ownerBinding, privateReaderBinding]}),
+    publicPolicy:
+        new PolicyConstructor({bindings: [ownerBinding, allReadBinding]}),
+    publicPolicyWithReader: new PolicyConstructor(
+        {bindings: [ownerBinding, allReadWithReaderBinding]},
+        ),
+  };
+}
+
+it('Does not try to make score assets readable when not gd user', () => {
+  cy.get('@gapiEmail')
+      .then((emailStub) => emailStub.returns('some-user@givedirectly.org'));
+  const getPolicyStub = cy.stub(ee.data, 'getIamPolicy');
+  waitForTrackAndAssertNormalStubs();
+  // Give other thread a chance to call listAssets if that's going to happen.
+  cy.wait(0).then(() => expect(getPolicyStub).to.not.be.called);
```

Comment:
cyQueue here? Is cy.wait(0) and cy.wrap(null) slightly different?

Reply:
Yes, they're slightly different. See https://github.com/givedirectly/Google-Partnership/blob/master/cypress/integration/unit_tests/cyqueue_test.js#L18 for a hint of the behavior. `cy.wrap()` and `cy.get()` do not explicitly give up control to another thread: if the object inside the wrap is an ordinary one, or if the element is findable on the page, respectively, they continue execution in the current thread. That means that if another thread is waiting to go, it won't get a chance. `cy.wait()` will always give up control: `cy.wait(0)` puts itself right back on the event loop stack, but will allow anything else that is waiting to run to go first.

That's why, in the test I linked to, even though the Promise had resolved before both of them, the `.then` clause for `cy.get()` ran before the Promise's `.then` clause, while the one for `cy.wait()` ran after.